### ISSUE TITLE
Linux clients freeze irregularly

### ISSUE THREAD

**Issue by ETLegacyTeam** [etlegacy/etlegacy#401]
Author: Unode (Unode)
Date: 2013-11-15
Redmine Issue: 401, https://dev.etlegacy.com/issues/401

---

I was playing with a few people on a LAN server running on Ubuntu precise.
We were playing single map (oasis) with 10 bots in each team for a total of 20 players active (bots would leave/join if humans leave/join)
Every now and then several players in the same team would have their clients freeze and loop the last second of sound.

Since this seemed to affect multiple players simultaneously it seemed something triggered on the server side.
At least once the players that froze were in the same fireteam, I cannot confirm this information for the other crashes.

On the client side there's no error but the client has to be killed/stopped manually.

On the server the output includes the following error message (some whitespace between lines removed):

```
getdot failed.
 
scripts/goals/goal_fireteam.gm(288) :   if ( entPos.IsZero() )
 
callstack..

288: Follow
137: Update

```

which repeats several times until a client reconnects (server is setup to freeze bots on idle).

In addition to this I also noticed that the locations of players/bots in the fireteam showed up as "unknown" instead of the actual location in the map.

*Edit:*
Below is included a list of the situations where the client froze and where it didn't.

Freezing:
* 32bit 2.72 client connecting to a local dedicated server running *with* bots. (legacy mod + omnibot)
* 32bit 2.72 client connecting to test server @ etlegacy.com. (legacy mod + omnibot)
* 32bit 2.72 client connecting to a non-dedicated server on Windows running *with* bots (legacy mod + omnibot)

Not Freezing:
* 32bit git-00949de client connecting to a local dedicated server running *with* bots. (legacy mod + omnibot). Over 2hours of playtime.
* 32bit 2.72 client connecting to a local dedicated server running *without* bots. (legacy mod + omnibot). Playing solo, running and shooting around and doing objectives.
* 32bit 2.72 client connecting to a local dedicated server running *without* bots. (legacy mod + omnibot). Idling for hours.
* 64bit 2.72 client connecting to a local dedicated server running *with* bots. (legacy mod + omnibot). Over 4hour of playtime.
* 64bit git-3cddea4 client connecting to a local dedicated server running *with* bots. (legacy mod + omnibot). Over 1hour of playtime.
* 32bit 2.72 client connecting to an online server running Silent mod *with* bots (silent mod + bots (engine?)). Some 30mins playtime. *(requires longer testing)*

**Comment by ETLegacyTeam** [comment]
From: Unode (Unode)
Date: 2013-11-15T21:30:20Z

---

Some more output that happened on the server while I was testing the "unknown" location in the fireteam, only this time the client didn't freeze.

```
...

expecting param 0 as vec3, got null

 
scripts/goals/goal_fireteam.gm(265) :   if(this.Goto(this.TargetPosition, 128.f) == EVENT.PATH_SUCCESS)

 
callstack..

265: DynamicGoto

...

getdot failed.

 
scripts/goals/goal_fireteam.gm(283) :   if( this.TargetPosition.IsZero() )

 
callstack..

283: Follow

137: Update

...

getdot failed.

 
scripts/goals/goal_fireteam.gm(288) :   if ( entPos.IsZero() )

 
callstack..

288: Follow

137: Update


```

**Comment by ETLegacyTeam** [comment]
From: crapshoot (crapshoot)
Date: 2013-11-15T23:36:51Z

---

This is an omni-bot script issue. You should disable the fireteam goal as it's in an experimental state and is no longer being maintained. Issue the /bot ft_controls 0 command in console to disable it.

**Comment by ETLegacyTeam** [comment]
From: @IR4T4 (IR4T4)
Date: 2013-11-16T07:19:47Z

---

I can confirm this but I'm not sure if it's related to bots. We have had some LAN games w/o bots on Debian machines. From time to time some clients (as far as I do remember they were in the same team) did use 100% CPU.

**Comment by ETLegacyTeam** [comment]
From: @IR4T4 (IR4T4)
Date: 2013-12-27T10:42:55Z

---

My client doesn't crash anymore ... I think this issue was related to WP_NONE usage (which is fixed).

**Comment by ETLegacyTeam** [comment]
From: 3togo (3togo)
Date: 2013-12-28T22:03:54Z

---

IR4T4 wrote:
> My client doesn't crash anymore ... I think this issue was related to WP_NONE usage (which is fixed).

In my case, the cpu as shown by "TOP" surged to beyond 120% and then crashed randomly.  Is there any chances to curb the surge of CPU hog

**Comment by ETLegacyTeam** [comment]
From: @IR4T4 (IR4T4)
Date: 2013-12-29T01:45:37Z

---

3togo wrote:
> In my case, the cpu as shown by "TOP" surged to beyond 120% and then crashed randomly.  Is there any chances to curb the surge of CPU hog

Which ETL version is used?

**Comment by ETLegacyTeam** [comment]
From: 3togo (3togo)
Date: 2013-12-29T20:01:20Z

---

IR4T4 wrote:
> 3togo wrote:
> > In my case, the cpu as shown by "TOP" surged to beyond 120% and then crashed randomly.  Is there any chances to curb the surge of CPU hog
> 
> Which ETL version is used?

etlegacy-linux-2.71rc3-full-installer.sh

**Comment by ETLegacyTeam** [comment]
From: @IR4T4 (IR4T4)
Date: 2014-01-17T12:38:59Z

---

Works for me.

**Comment by ETLegacyTeam** [comment]
From: Unode (Unode)
Date: 2014-01-17T13:38:16Z

---

Thanks IR4T4, I'll make sure I'll give it a test once rc4 comes out and update then.

**Comment by ETLegacyTeam** [comment]
From: Unode (Unode)
Date: 2015-03-22T03:39:25Z

---

Trying to reopen an old ticket.

I've re-experienced the same problem from before but now I no longer see the console errors related with omnibot.

The situation happens on a linux 64bit system running the 2.72 32bit release.
I was able to reproduce the freeze while connecting to the test-server running at etlegacy.com and with a local dedicated server.
The time for the client to freeze varies but was around 30 minutes on the etlegacy.com server.

I tried to compile the latest version from git. I was able to compile the 64bit version which after a few hours of testing didn't freeze.
However this was a debug build so I'm not 100% sure if this is fixed in latest git or if not being an optimized build hides the problem.

RaFaL and spyhawk provided some guidance on IRC, including build scripts.
I was not able to compile a 32bit version due to several missing dependencies.
RaFaL was kind enough to provide one using latest git but the binaries were incompatible with my system causing a segfault immediately at execution.

So with that, I'd like to say this is indeed fixed but I'd like to try it under the same conditions, i.e. a 32bit build with same parametrization used for release of 2.72.
If you can provide this build I'd be happy to test it.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2015-03-22T09:37:35Z

---

RaFal build didn't work for me either. You should try a release build of latest master (even 64 bit) and see if you can reproduce.

**Comment by ETLegacyTeam** [comment]
From: Unode (Unode)
Date: 2015-03-23T02:00:35Z

---

So reporting my attempts...

With 64bit builds I cannot reproduce the freeze. Tried both build from source and the official 2.72 release.
Tried the official 2.72 32bit build and it froze. Also tried updating omnibots to the latest available and it still froze.
Still with the 32bit build, now playing without bots, I got bored after 30 mins running around shooting walls. During this period it didn't freeze.
I'm going to try to leave the client running on its own and see if it freezes. I'll update if it does.

So summing up, it's specific to 32bits and maybe related/triggered by having bots around.
The 32bit build with debug symbols will come in handy to try to figure this out.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2015-03-23T12:22:31Z

---

Sounds like #746.

Edit: sorry, did miss this was a client issue.

**Comment by ETLegacyTeam** [comment]
From: Unode (Unode)
Date: 2015-03-23T12:51:02Z

---

Well I've never had server freezes, only client. When the client freezes I can restart the client and reconnect to the same server without problems.
One more bit of info, if I run the client in window mode and launch it from a terminal I can ctrl+c the process and it exits "cleanly". I.e. next time etl starts it won't complain that it crashed before.

**Comment by ETLegacyTeam** [comment]
From: Unode (Unode)
Date: 2015-03-24T11:29:39Z

---

One more update. So leaving the client running for a few hours without input (and disabled timeouts to prevent kicking from server) it didn't freeze.
I also tried to play other mods by joining online servers and for the time I played, which was short, no freeze happened. This was with Silent mod.

I'll edit the main post to get the information together in one place.

**Comment by ETLegacyTeam** [comment]
From: @IR4T4 (IR4T4)
Date: 2015-03-24T16:05:17Z

---

I think this is a mod issue - it's just odd omnibot has issues with the position of entities. I do assume you did enable antilag. What happens if g_antilag is set to 0?

**Comment by ETLegacyTeam** [comment]
From: Unode (Unode)
Date: 2015-03-24T23:30:56Z

---

Same thing, it still crashes.
g_antilag 0 set on the server and client, confirmed on the console when map was running.

**Comment by ETLegacyTeam** [comment]
From: Unode (Unode)
Date: 2015-03-25T11:06:16Z

---

So I was finally able to compile a 32bit build from git 00949de and I wasn't able to reproduce the freeze.
I still need to do further testing to make sure cause this build ended up with no sound.
I used the easybuild script and didn't see any missing library but ... perhaps I'm still missing something.

**Comment by ETLegacyTeam** [comment]
From: Unode (Unode)
Date: 2015-03-26T16:17:15Z

---

So confirming the previous comment, sounds problems were due to missing alsa during compilation.
After this all went well and I cannot reproduce the freeze with latest git-head in 32bits either.

I'll leave this ticket open in feedback state. Once the first 2.73 candidate comes out I'll do further testing and mark this a fixed if no longer reproducible.

Thanks for the help

**Comment by ETLegacyTeam** [comment]
From: @IR4T4 (IR4T4)
Date: 2015-03-26T18:23:18Z

---

Let's close it - we can open this issue again IF the bug occures again.
